core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -3277,6 +3346,145 @@ private void optimiseJoins(QueryModel model) throws SqlException {
         }
     }
 
+    /**
+     * Optimises models with ASOF joins, to push up .
+     * <p>
+     * This method traverses the query model and optimises it by restructuring
+     * the nested models to facilitate ASOF joins.
+     *
+     * @param model the query model to optimise
+     */
+    private void optimiseModelsWithASOFJoins(SqlExecutionContext executionContext, QueryModel model) throws SqlException {
+        if (model == null)
+            return;
+
+        ObjList<QueryModel> joinModels = model.getJoinModels();
+        for (int i = 1, n = joinModels.size(); i < n; i++) {
+            optimiseModelsWithASOFJoins(executionContext, joinModels.getQuick(i));
+        }
+
+        QueryModel temporayModel = model;
+        while (temporayModel != null) {
+            optimiseModelsWithASOFJoins(executionContext, temporayModel.getNestedModel());
+            optimiseModelsWithASOFJoins(executionContext, model.getUnionModel());
+            temporayModel = temporayModel.getNestedModel();
+        }
+
+        QueryModel targetModel = model.getNestedModel();
+        QueryModel parentModel = model;
+        /*
+         target-model is the model containing asof join, where-clause, limit-clause, order-by-clause.
+         parent-model is the parent of target-model
+         */
+        while (targetModel != null && targetModel.getJoinModels().size() != 2) {
+            parentModel = targetModel;
+            targetModel = targetModel.getNestedModel();
+        }
+
+        if (!isModelEligibleForASOFJoinOptimisation(executionContext, targetModel, parentModel))
+            return;
+
+        QueryModel baseTableModel = model;
+        while (baseTableModel.getNestedModel() != null) {
+            baseTableModel = baseTableModel.getNestedModel();
+        }
+
+        /*
+          level-0 model which contains everything inside target model
+        */
+        QueryModel level0 = targetModel.getNestedModel();
+        QueryModel copyModel = level0 == null ? targetModel : level0;
+
+        /*
+          nested of target model is null, construct level-0 using base table
+         */
+        if (level0 == null) {
+            level0 = QueryModel.FACTORY.newInstance();
+            level0.setTimestamp(baseTableModel.getTimestamp());
+            level0.copyColumnsFrom(copyModel, queryColumnPool, expressionNodePool);
+            //if target model has both table name and alias.
+            level0.setTableNameExpr(copyModel.getTableNameExpr() == null ? null : copyModel.getTableNameExpr());
```

Comment:
`copyModel.getTableNameExpr() == null ? null : copyModel.getTableNameExpr()` could be just `copyModel.getTableNameExpr()`. Am I missing something?